pubsublite-spark-sql-streaming/src/main/java/com/google/cloud/pubsublite/spark/PslDataSourceOptions.java

```diff
@@ -0,0 +1,108 @@
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsublite.spark;
+
+import com.google.auto.value.AutoValue;
+import com.google.cloud.pubsublite.SubscriptionPath;
+import java.io.Serializable;
+import java.util.Optional;
+import javax.annotation.Nullable;
+import org.apache.spark.sql.sources.v2.DataSourceOptions;
+
+@AutoValue
+public abstract class PslDataSourceOptions implements Serializable {
+
+  @Nullable
+  public abstract String credentialsAccessToken();
+
+  @Nullable
+  public abstract String credentialsKey();
+
+  @Nullable
+  public abstract String credentialsFile();
+
+  public abstract SubscriptionPath subscriptionPath();
+
+  public abstract long maxBytesOutstanding();
+
+  public abstract long maxMessagesOutstanding();
+
+  public abstract boolean consumeFromHead();
+
+  public abstract long maxBatchOffsetRange();
+
+  public static Builder builder() {
+    return new AutoValue_PslDataSourceOptions.Builder()
+        .credentialsAccessToken(null)
+        .credentialsKey(null)
+        .credentialsFile(null)
+        .maxBytesOutstanding(Constants.DEFAULT_BYTES_OUTSTANDING)
+        .maxMessagesOutstanding(Constants.DEFAULT_MESSAGES_OUTSTANDING)
+        .maxBatchOffsetRange(Constants.DEFAULT_BATCH_OFFSET_RANGE)
+        .consumeFromHead(true);
+  }
+
+  public static PslDataSourceOptions fromSparkDataSourceOptions(DataSourceOptions options) {
+    if (!options.get(Constants.SUBSCRIPTION_CONFIG_KEY).isPresent()) {
+      throw new IllegalArgumentException(Constants.SUBSCRIPTION_CONFIG_KEY + " is required.");
+    }
+
+    Builder builder = builder();
+    Optional<String> value;
+    if ((value = options.get(Constants.CREDENTIALS_ACCESS_TOKEN_CONFIG_KEY)).isPresent()) {
+      builder.credentialsAccessToken(value.get());
+    } else if ((value = options.get(Constants.CREDENTIALS_KEY_CONFIG_KEY)).isPresent()) {
+      builder.credentialsKey(value.get());
+    } else if ((value = options.get(Constants.CREDENTIALS_FILE_CONFIG_KEY)).isPresent()) {
+      builder.credentialsFile(value.get());
+    }
+    builder.subscriptionPath(
+        SubscriptionPath.parse(options.get(Constants.SUBSCRIPTION_CONFIG_KEY).get()));
+    builder.maxBytesOutstanding(
+        options.getLong(
+            Constants.BYTES_OUTSTANDING_CONFIG_KEY, Constants.DEFAULT_BYTES_OUTSTANDING));
+    builder.maxMessagesOutstanding(
+        options.getLong(
+            Constants.MESSAGES_OUTSTANDING_CONFIG_KEY, Constants.DEFAULT_MESSAGES_OUTSTANDING));
+    builder.consumeFromHead(options.getBoolean(Constants.CONSUME_FROM_HEAD_CONFIG_KEY, true));
+    builder.maxBatchOffsetRange(
```

Comment:
The default value (as I addressed in the previous comment) would be large, and most of the time user should not even bother about this setting. The only case user would care is when having large backlog and consume from commit instead of head. If for example they run some complicated aggregation that takes time, they might want to lower this number to make sure they don't lose much progress if random issues happen.
I will document ^ clearly in the documentation.

Reply:
The user should never care about this setting. IIUC, batches will be capped at the head offset when near head, and we want to read as much data as possible when in backlog. Does this affect checkpointing of later steps at all? I thought it only affected the loading phase?

Please remove this setting until it is requested.

Reply:
Discussed offline, keeping this as a constant but not user configurable. Added a TODO to revisit for performance tuning later.